### DISCUSSION TITLE
Mitutoyo Digital Scale Serial Interfacing with Wiznet 5500EVB Pico using Arduino IDE

### DISCUSSION THREAD

**Discussion by themadgreek** [earlephilhower/arduino-pico#1169] [Q&A] [answered]
Hi 

I am currently using a Mitutoyo digital scale via serial interface thru an Arduino UNO to read distance measurements, Thanks to sspence from
[Sspence's Projects - Instructables]
(https://www.instructables.com/member/sspence/instructables/)

I am hoping to have this setup with the Wiznet 5500 EVB Pico instead of the UNO using Arduino IDE. I got the code I had for the uno to be compiled and uploaded to the Pico, but it doesn't seem to work at all. Hoping you would have some insights on this matter.
thanks!!

This is the code I have from sspence's page.

`int req = 5; //mic REQ line goes to pin 5 through q1 (Arduino high pulls request line low)
int dat = 2; //mic Data line goes to pin 2
int clk = 3; //mic Clock line goes to pin 3
int i = 0;
int j = 0;
int k = 0;
int signCh = 8;
int sign = 0;
int decimal;
float dpp;
int units;

byte mydata[14];
String value_str;
long value_int; //was an int, could not measure over 32mm
float value;

void setup() {

Serial.begin(9600);

pinMode(req, OUTPUT);

pinMode(clk, INPUT_PULLUP);

pinMode(dat, INPUT_PULLUP);

digitalWrite(req,LOW); // set request at high

}

void loop() {

digitalWrite(req, HIGH); // generate set request
for( i = 0; i < 13; i++ ) {
k = 0;
for (j = 0; j < 4; j++) {
while( digitalRead(clk) == LOW) {
} // hold until clock is high
while( digitalRead(clk) == HIGH) {
} // hold until clock is low
bitWrite(k, j, (digitalRead(dat) & 0x1));
}

mydata[i] = k;

}
sign = mydata[4];
value_str = String(mydata[5]) + String(mydata[6]) + String(mydata[7]) + String(mydata[8] + String(mydata[9] + String(mydata[10]))) ;
decimal = mydata[11];
units = mydata[12];

value_int = value_str.toInt();
if (decimal == 0) dpp = 1.0;
if (decimal == 1) dpp = 10.0;
if (decimal == 2) dpp = 100.0;
if (decimal == 3) dpp = 1000.0;
if (decimal == 4) dpp = 10000.0;
if (decimal == 5) dpp = 100000.0;

value = value_int / dpp;

if (sign == 0) {
Serial.println(value,decimal);
}
if (sign == 8) {
Serial.print("-"); Serial.println(value,decimal);
}

digitalWrite(req,LOW);
delay(100);
}`

![mitutoyo 1](https://user-images.githubusercontent.com/125086824/218184732-23abaa1d-9fb5-4492-95b4-66a0086b12c2.png)
![mitutoyo 2](https://user-images.githubusercontent.com/125086824/218184767-681ed2fc-dc0e-44c4-b0a8-b72c3cb0c200.png)
<img width="506" alt="PICO Pin Out" src="https://user-images.githubusercontent.com/125086824/218184814-68347125-f518-427a-be12-1202cfcd3a8e.PNG">

**Comment by maxgerhardt** [comment, marked as answer]
Please format the cpp code correctly, see [here](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting)

>but it doesn't seem to work at all. 

Not enough info, what's the serial output, what exactly doesn't seem to work? No output at all, wrong output?

Onto which pins did you connect the circuit? Whe you use
```cpp
int req = 5; //mic REQ line goes to pin 5 through q1 (Arduino high pulls request line low)
int dat = 2; //mic Data line goes to pin 2
int clk = 3; //mic Clock line goes to pin 3
``` 
It's expected to be connected to GP5, GP2, GP3.

**Reply by themadgreek** [reply]
Thanks @maxgerhardt

I had it connected to the wrong pins. I am very new with Picos and the pin # and GP# kinda got me confused. Your answer was right on the money!!

**Comment by themadgreek** [comment]
Hi @maxgerhardt 

Im still a basic beginner with the Pico. How can I send the data acquired via ethernet? I want the data accessed remotely from a computer through the network.

**Reply by maxgerhardt** [reply]
Have you already found + tried Arduino     ethernet sketches for the WIznet 5500EVB board?

**Reply by themadgreek** [reply]
Yes i have been trying to, but I'm not sure how to just send the data acquired from the above code via ethernet.

**Reply by themadgreek** [reply]
All the code above does is that it takes the digital micrometer (Mitutoyo) distance measurement and have it accessible in serial. I was hoping to utilize the ethernet connection with the Wiznet w5500 EVB pico to send the measurement data to the network.

**Reply by maxgerhardt** [reply]
Do you already have a sketch that starts the ethernet chip and acquires a dynamic IP address? Beyond that it's just using the EthernetClient (TPC) or the UDP.h library. Of course you gotta setup an apprpriate server / endpoint for the Pico to send the data to, too.

**Reply by themadgreek** [reply]
No, I don't have a sketch yet that does that.

**Reply by maxgerhardt** [reply]
Well you know onto which pins the W5500 is connected per the diagram above 

![image](https://user-images.githubusercontent.com/26485477/219419832-ebeff1aa-356e-4caf-8018-3bfa9d7f884b.png)

So the first thought would be to grab the official Arduino Ethernet library at https://github.com/arduino-libraries/Ethernet/, grab the WebClient/WebClient.ino example, give it the right CS pin (17) and see what it does. You may or may not also have to drive the RSTn pin (20) high before you do that.

**Reply by themadgreek** [reply]
When you meant "give it the right CS pin (17) and see what it does" , did you mean to assign the data line pin to GP17 instaed of the original GP2?

**Reply by maxgerhardt** [reply]
You tell the library what your CS pin is, not the other way around. You wouldn't wanna modify your hardware anyways. So just use a `Ethernet.init(17);` statement in the setup() of the sketch where all these comments are.

Where do you see "the original GP2" mentioned in the sketch?

**Reply by themadgreek** [reply]
I meant the original GP2 with reference to the code i have to iterface the data off the MITutoyo

**Reply by maxgerhardt** [reply]
Why would that need changing? The hardware connects to your digital scale are separate from the on-board data connection to the Ethernet chip (aka SPI to the W5500). These are two separate circuits.

**Reply by themadgreek** [reply]
I understand the two circuits are different, But where i am confused is with the code. Do i just insert Ethernet.init(17);  and the Ip address to the code?

**Reply by maxgerhardt** [reply]
You should first try a sketch that *only* uses Ethernet and ignores your other circuit. You'll figure out how to merge / integrate your old sketch with the Ethernet code later. As I said, try https://github.com/arduino-libraries/Ethernet/blob/master/examples/WebClient/WebClient.ino first with a `Ethernet.init(17);` and possibly a `pinMode(20, OUTPUT); digitalWrite(20, HIGH);` to release the reset line.

**Reply by themadgreek** [reply]
ok perfect, that makes sense Sorry I'm new to this and coding is not my forte. Figuring this out this interesting area slowly

**Reply by themadgreek** [reply]
Hi @maxgerhardt 
I tried the web client example, but it is not working on the system that i was trying. And i tried the chat server example and that seem to work fine. Do you think i can use the chat server code to modify the digital scale code to have the functionality i was hoping for?

**Comment by themadgreek** [comment]
Hi @maxgerhardt 
I figured out how to send the distance data off the digital scale via ethernet. This is the code I ended up figuring out. The one thing I've noticed, and I haven't sorted out yet is that to receive the data via ethernet the connection has to see some kinda handshake or message. So, in my case the serial monitor would bring up distance data right away without doing anything, but data via ethernet only comes in when I send a character or string to it, and then the data just keeps coming in just the way it's supposed to. I realize it's something in my code that does this, but I couldn't find out what is doing this yet. But the code works to an extent.

`#include <SPI.h>
#include <Ethernet.h>

byte mac[] = { 0x90, 0xA2, 0xDA, 0x0D, 0xA3, 0x12 };
IPAddress ip(10,10,10,112);
byte subnet[] = { 255, 255, 255, 0 };

EthernetServer server(23);

int req = 5; //mic REQ line goes to pin 5 through q1 (Arduino high pulls request line low)
int dat = 2; //mic Data line goes to pin 2
int clk = 3; //mic Clock line goes to pin 3
int i = 0;
int j = 0;
int k = 0;
int signCh = 8;
int sign = 0;
int decimal;
float dpp;
int units;

byte mydata[14];
String value_str;
long value_int; //was an int, could not measure over 32mm
float value;

void setup() {
  
Ethernet.init(17);  // WIZnet W5100S-EVB-Pico W5500-EVB-Pico

Serial.begin(9600);

pinMode(req, OUTPUT);

pinMode(clk, INPUT_PULLUP);

pinMode(dat, INPUT_PULLUP);

pinMode(20, OUTPUT); 

digitalWrite(20, HIGH); //change

digitalWrite(req,LOW); // set request at high

Ethernet.begin(mac, ip, subnet);   
  
server.begin();


delay (100);
}

void loop() {

EthernetClient client = server.available();

digitalWrite(req, HIGH); // generate set request

for( i = 0; i < 13; i++ ) {
k = 0;
for (j = 0; j < 4; j++) {
while( digitalRead(clk) == LOW) {
} // hold until clock is high
while( digitalRead(clk) == HIGH) {
} // hold until clock is low
bitWrite(k, j, (digitalRead(dat) & 0x1));
}

mydata[i] = k;

}
sign = mydata[4];
value_str = String(mydata[5]) + String(mydata[6]) + String(mydata[7]) + String(mydata[8] + String(mydata[9] + String(mydata[10]))) ;
decimal = mydata[11];
units = mydata[12];

value_int = value_str.toInt();
if (decimal == 0) dpp = 1.0;
if (decimal == 1) dpp = 10.0;
if (decimal == 2) dpp = 100.0;
if (decimal == 3) dpp = 1000.0;
if (decimal == 4) dpp = 10000.0;
if (decimal == 5) dpp = 100000.0;

value = value_int / dpp;

if (sign == 0) {
Serial.println(value,decimal);
client.println(value,decimal);
}
if (sign == 8) {
Serial.print("-"); 
client.print(", ");
Serial.println(value,decimal);
client.println(value,decimal);

}

digitalWrite(req,LOW);
delay(100);


//EthernetClient client = server.available();
// if (client) {
//   while (client.connected()) {
//     client.println(value,decimal);
//     client.print(", ");
//     client.println(value,decimal);
     
// }
//client.stop();
//}

}`

**Reply by maxgerhardt** [reply]
I think this is due to your usage of `server.available()`. If you look at the function code, it not only requires that a socket is in the `SnSR::ESTABLISHED` state, but **also** that that client already has sent some data which is available for reading (`if (Ethernet.socketRecvAvailable(i) > 0)`).

See [source code](https://github.com/arduino-libraries/Ethernet/blob/master/src/EthernetServer.cpp#L40-L105).

If you were to use `server.accept();`  that problem should go away. But then you also need to take special care to save the returned EthernetClient object because it will not be returned by the function again. Check out the techniques from https://github.com/arduino-libraries/Ethernet/blob/master/examples/AdvancedChatServer/AdvancedChatServer.ino#L79-L93.

**Reply by themadgreek** [reply]
I tried it, now it brings the first data and stops, without having to send a character, no other data coming in thereafter, while serial monitor brings in all the stream of data always.

**Reply by maxgerhardt** [reply]
Updated code, [correctly formatted](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) please?

**Reply by themadgreek** [reply]
Didnt change much, just changed server.available(); to server.accept();
And i ran into the issue you warned about. 

`#include <SPI.h>
#include <Ethernet.h>

byte mac[] = { 0x90, 0xA2, 0xDA, 0x0D, 0xA3, 0x12 };
IPAddress ip(10,10,10,112);
byte subnet[] = { 255, 255, 255, 0 };

EthernetServer server(23);

int req = 5; //mic REQ line goes to pin 5 through q1 (Arduino high pulls request line low)
int dat = 2; //mic Data line goes to pin 2
int clk = 3; //mic Clock line goes to pin 3
int i = 0;
int j = 0;
int k = 0;
int signCh = 8;
int sign = 0;
int decimal;
float dpp;
int units;

byte mydata[14];
String value_str;
long value_int; //was an int, could not measure over 32mm
float value;

void setup() {
  
Ethernet.init(17);  // WIZnet W5100S-EVB-Pico W5500-EVB-Pico

Serial.begin(9600);

pinMode(req, OUTPUT);

pinMode(clk, INPUT_PULLUP);

pinMode(dat, INPUT_PULLUP);

pinMode(20, OUTPUT); 

digitalWrite(20, HIGH); //change

digitalWrite(req,LOW); // set request at high

Ethernet.begin(mac, ip, subnet);   
  
server.begin();


delay (100);
}

void loop() {

EthernetClient client = server.available();

digitalWrite(req, HIGH); // generate set request

for( i = 0; i < 13; i++ ) {
k = 0;
for (j = 0; j < 4; j++) {
while( digitalRead(clk) == LOW) {
} // hold until clock is high
while( digitalRead(clk) == HIGH) {
} // hold until clock is low
bitWrite(k, j, (digitalRead(dat) & 0x1));
}

mydata[i] = k;

}
sign = mydata[4];
value_str = String(mydata[5]) + String(mydata[6]) + String(mydata[7]) + String(mydata[8] + String(mydata[9] + String(mydata[10]))) ;
decimal = mydata[11];
units = mydata[12];

value_int = value_str.toInt();
if (decimal == 0) dpp = 1.0;
if (decimal == 1) dpp = 10.0;
if (decimal == 2) dpp = 100.0;
if (decimal == 3) dpp = 1000.0;
if (decimal == 4) dpp = 10000.0;
if (decimal == 5) dpp = 100000.0;

value = value_int / dpp;

if (sign == 0) {
Serial.println(value,decimal);
client.println(value,decimal);
}
if (sign == 8) {
Serial.print("-"); 
client.print(", ");
Serial.println(value,decimal);
client.println(value,decimal);

}

digitalWrite(req,LOW);
delay(100);


//EthernetClient client = server.accept();
// if (client) {
//   while (client.connected()) {
//     client.println(value,decimal);
//     client.print(", ");
//     client.println(value,decimal);
     
// }
//client.stop();
//}

}`

**Reply by maxgerhardt** [reply]
You are not managing the connected clients like in the example I linked you to, so this only works for one loop() iteration. Explicitly, you are missing the saving the accepted clients into an array, clearing them when disconnected, and iterating through all connected clients when you want to send data.

Check out https://gist.github.com/maxgerhardt/a33206a8e76ee6c0b66d574c4a1ad1e8. Untested of course.

**Reply by themadgreek** [reply]
Hi @maxgerhardt 

I tried the sketch you put together, and i works perfectly just how i wanted it to. Thanks a lot for helping, Now i can utilize the sketch as reference for any future builds. I really appreciate your help and your guidance!!

**Comment by SSpence** [comment]
Remove me from this email group immediately. I did not ask to subscribe to
this.

On Fri, Feb 17, 2023 at 5:15 PM Maximilian Gerhardt <
***@***.***> wrote:

> You are not managing the connected clients like in the example I linked
> you to, so this only works for one loop() iteration. Explicitly, you are
> missing the saving the accepted clients into an array, clearing them when
> disconnected, and iterating through all connected clients when you want to
> send data.
>
> Check out
> https://gist.github.com/maxgerhardt/a33206a8e76ee6c0b66d574c4a1ad1e8.
> Untested of course.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/earlephilhower/arduino-pico/discussions/1169#discussioncomment-5015082>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIPHXMDA4BBKMYXCZXK7ELWX72AFANCNFSM6AAAAAAUYGPCQI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> .com>
>
-- 
Shala Spence, M.Ed.
NEAAAT Instructional Coach
2018-19 NEAAAT Teacher of the Year
FIRST Robotics Competition Team Lead Coach
#teamNEAAAT

-- 










*E-mail correspondence to and from this address is subject to 
North Carolina's Public Records Act,  N.C. Gen. Stat. Sec. 132, and may be 
disclosed to third parties.  However, Federal and State law protects 
personal health and other information that may be contained in this e-mail 
from unauthorized disclosure.  If you are not the intended recipient, 
please delete this e-mail and any accompanying documents and contact the 
sender immediately.  Unauthorized disclosure, copying or distribution of 
any confidential or privileged content of this e-mail is prohibited.*

**Reply by maxgerhardt** [reply]
Idk how you got into this thread, since I didn't even quote your username or something, but just use the unsubscribe link that Github gave you

**Comment by SSpence** [comment]
I do not have a GitHub account.

On Fri, Feb 17, 2023 at 5:19 PM Maximilian Gerhardt <
***@***.***> wrote:

> Idk how you got into this thread, since I didn't even quote your username
> or something, but just use the unsubscribe link that Github gave you
>
> or unsubscribe
>
> https://github.com/notifications/unsubscribe-auth/AFIPHXMDA4BBKMYXCZXK7ELWX72AFANCNFSM6AAAAAAUYGPCQI
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/earlephilhower/arduino-pico/discussions/1169#discussioncomment-5015117>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIPHXLVAZRQXVLVH3JSFKDWX72P5ANCNFSM6AAAAAAUYGPCQI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> .com>
>
-- 
Shala Spence, M.Ed.
NEAAAT Instructional Coach
2018-19 NEAAAT Teacher of the Year
FIRST Robotics Competition Team Lead Coach
#teamNEAAAT

-- 










*E-mail correspondence to and from this address is subject to 
North Carolina's Public Records Act,  N.C. Gen. Stat. Sec. 132, and may be 
disclosed to third parties.  However, Federal and State law protects 
personal health and other information that may be contained in this e-mail 
from unauthorized disclosure.  If you are not the intended recipient, 
please delete this e-mail and any accompanying documents and contact the 
sender immediately.  Unauthorized disclosure, copying or distribution of 
any confidential or privileged content of this e-mail is prohibited.*

**Reply by maxgerhardt** [reply]
Okay so https://github.com/SSpence/FRC7715 is not a repository of yours.. under your username?

Well in any case, looks like github is glitching, if it keeps happening contact support. I don't have the power to unsubscribe you.
